src/main/PersistentState.cpp

```diff
@@ -36,47 +38,94 @@ PersistentState::dropAll(Database& db)
     st.execute(true);
 }
 
-string
+std::string
 PersistentState::getStoreStateName(PersistentState::Entry n)
 {
     if (n < 0 || n >= kLastEntry)
     {
         throw out_of_range("unknown entry");
     }
-    return mapping[n];
+    auto res = mapping[n];
+    if (n == kLastSCPData)
+    {
+        if (mSCPStateCounter < 0)
+        {
+            throw std::runtime_error(
+                "Previous SCP state must be properly restored");
+        }
+        res += std::to_string(mSCPStateCounter);
+    }
+    return res;
 }
 
-string
+std::string
 PersistentState::getState(PersistentState::Entry entry)
 {
-    string res;
+    return getFromDb(entry);
+}
 
-    string sn(getStoreStateName(entry));
+void
+PersistentState::setState(PersistentState::Entry entry,
+                          std::string const& value)
+{
+    updateDb(entry, value);
+}
 
-    auto& db = mApp.getDatabase();
-    auto prep = db.getPreparedStatement(
-        "SELECT state FROM storestate WHERE statename = :n;");
-    auto& st = prep.statement();
-    st.exchange(soci::into(res));
-    st.exchange(soci::use(sn));
-    st.define_and_bind();
+std::vector<std::string>
+PersistentState::getSCPState()
+{
+    // First, collect all slots persisted
+    std::vector<std::string> states;
+    mSCPStateCounter = 0;
+    for (int i = 0; i <= Herder::MAX_SLOTS_TO_REMEMBER; i++)
     {
-        auto timer = db.getSelectTimer("state");
-        st.execute(true);
+        auto val = getFromDb(kLastSCPData);
```

Comment:
I think those helper functions `updateDb/getFromDb` are better off taking the key name (string) as parameter

Reply:
Updated.